components/tidb_query/src/rpn_expr/impl_time.rs

```diff
@@ -347,6 +371,106 @@ mod tests {
         }
     }
 
+    #[test]
+    fn test_date_diff() {
+        use crate::expr::{EvalConfig, EvalContext, Flag, SqlMode};
+        use std::sync::Arc;
+
+        let cases = vec![
+            (
+                "0000-01-01 00:00:00.000000",
+                "0000-01-01 00:00:00.000000",
+                0,
+            ),
+            (
+                "2018-02-01 00:00:00.000000",
+                "2018-02-01 00:00:00.000000",
+                0,
+            ),
+            (
+                "2018-02-02 00:00:00.000000",
+                "2018-02-01 00:00:00.000000",
+                1,
+            ),
+            (
+                "2018-02-01 00:00:00.000000",
+                "2018-02-02 00:00:00.000000",
+                -1,
+            ),
+            (
+                "2018-02-02 00:00:00.000000",
+                "2018-02-01 23:59:59.999999",
+                1,
+            ),
+            (
+                "2018-02-01 23:59:59.999999",
+                "2018-02-02 00:00:00.000000",
+                -1,
+            ),
+        ];
+
+        for (lhs, rhs, exp) in cases {
+            let mut ctx = EvalContext::default();
+            let output = RpnFnScalarEvaluator::new()
+                .push_param(Some(Time::parse_datetime(&mut ctx, lhs, 6, true).unwrap()))
+                .push_param(Some(Time::parse_datetime(&mut ctx, rhs, 6, true).unwrap()))
+                .evaluate(ScalarFuncSig::DateDiff)
+                .unwrap();
+            assert_eq!(output, Some(exp));
+        }
+
+        let invalid_zero_cases = vec![
+            (
+                Some("0000-00-00 23:59:59.999999"),
+                Some("2018-02-02 00:00:00.000000"),
+            ),
+            (
+                Some("2018-02-01 23:59:59.999999"),
+                Some("2018-00-00 23:59:59.999999"),
+            ),
+            (None, Some("2018-00-00 23:59:59.999999")),
+            (Some("2018-00-00 23:59:59.999999"), None),
+        ];
+
+        for (lhs, rhs) in invalid_zero_cases {
+            let mut ctx1 = EvalContext::default();
+            let output1 = RpnFnScalarEvaluator::new()
+                .push_param(
+                    lhs.map(|inner| Time::parse_datetime(&mut ctx1, inner, 6, true).unwrap()),
+                )
+                .push_param(
+                    rhs.map(|inner| Time::parse_datetime(&mut ctx1, inner, 6, true).unwrap()),
+                )
+                .evaluate::<Int>(ScalarFuncSig::DateDiff);
+            assert!(output1.is_err());
+
+            let mut cfg = EvalConfig::new();
+            cfg.set_flag(Flag::IN_UPDATE_OR_DELETE_STMT)
+                .set_sql_mode(SqlMode::NO_ZERO_DATE | SqlMode::STRICT_ALL_TABLES);
+
+            let mut ctx2 = EvalContext::new(Arc::new(cfg));
```

Comment:
It's better to use `ctx`. (There is no need to distinguish `ctx1` and `ctx2`, all of them can be `ctx`)

Reply:
Thanks, This is still WIP.  I' m going to refactor this part.

Reply:
> It's better to use `ctx`. (There is no need to distinguish `ctx1` and `ctx2`, all of them can be `ctx`)

Seems that non-vectorized has a bug.

Consider 
```sql
INSERT INTO `table` VALUES(datediff(NULL, '<invalid_zero_time>'));
```

it will retun NULL directly instead or error in `NO_ZERO_IN_DATE,STRICT_ALL_TABLES` mode. I will take a look.